CRM/Core/Smarty.php

```diff
@@ -176,15 +176,26 @@ public static function &singleton() {
    * @param bool $display
    *
    * @return bool|mixed|string
+   *
+   * @noinspection PhpDocMissingThrowsInspection
+   * @noinspection PhpUnhandledExceptionInspection
    */
   public function fetch($resource_name, $cache_id = NULL, $compile_id = NULL, $display = FALSE) {
     if (preg_match('/^(\s+)?string:/', $resource_name)) {
       $old_security = $this->security;
       $this->security = TRUE;
     }
-    $output = parent::fetch($resource_name, $cache_id, $compile_id, $display);
-    if (isset($old_security)) {
-      $this->security = $old_security;
+    try {
+      $output = parent::fetch($resource_name, $cache_id, $compile_id, $display);
+      if (isset($old_security)) {
+        $this->security = $old_security;
+      }
+    }
+    catch (Exception $e) {
+      if (isset($old_security)) {
+        $this->security = $old_security;
+      }
+      throw $e;
```

Comment:
Would this be equivalent/shorter as try/finally?

```php
try {
  $output = parent::fetch($resource_name, $cache_id, $compile_id, $display);
}
finally {
  if (isset($old_security)) {
    $this->security = $old_security;
  }
}
```

Reply:
yeah probably - I feel about the same about the two variants

Reply:
- the Smarty change is tangential to this PR - it solves the problem of tests falling over rather than individual tests failing - which plagued this work & other efforts in the past - but can be moved out